### PR TITLE
Synchronization into the develop branch after happy path end-to-end integration

### PR DESCRIPTION
**Pull request by gtorodelvalle**
- 100% tests passed
- 0 linting issues
- Assigned to @dmoranj 
- Feedback from @mrutid is more than welcome

### CODE REVIEW

File: lib/context_adapter_helper.js [hunk 1]
```diff
@@ -23,7 +23,9 @@
 
 'use strict';
 
+var caLogger = require('logops');
 var caConfig = require('./context_adapter_configuration');
+caLogger.setLevel(caConfig.LOGOPS_LEVEL);
```
Comment (dmoranj) [inline lib/context_adapter_helper.js]: Please, group all the requires in a single var statement, and don't mix code with declarations: move this statement to a line after all the var statements. 

On the other hand, setting the level in the root of a module may cause behaviors difficult to predict, as it is not easy to tell when this particular module will be loaded (as it will depend on the modules it depends from). I suggest to do it in an initialization function somewhere in the main file (or in a function you know when it will be executed, e.g.: the creation of the HTTP server).

File: lib/context_adapter_helper.js [hunk 2]
```diff
@@ -454,27 +535,41 @@ function getServiceDescriptor(request, operationDescriptor, callback) {
...
-          caConfig.SERVICE_ENTITY.AUTHENTICATION_ATTR_NAME,
-          caConfig.SERVICE_ENTITY.MAPPING_ATTR_NAME,
-          caConfig.SERVICE_ENTITY.TIMEOUT_ATTR_NAME
-        ]
+        isPattern: false
       }
+    ],
+    attributes: [
+      caConfig.SERVICE_ENTITY.INTERACTION_TYPE_ATTR_NAME,
+      caConfig.SERVICE_ENTITY.ENDPOINT_ATTR_NAME,
```
Comment (AlvaroVega) [inline lib/context_adapter_helper.js]: and PROVIDER_ATTR_NAME?

File: lib/context_adapter_helper.js [hunk 3]
```diff
@@ -205,6 +217,71 @@ function isAttributeIncluded(attributes, attributeName) {
 }
 
 /**
+ * Validates the attribute values received
+ * @param {object} context The context for logging
+ * @param {string} entityType The entity type (Button or Service)
+ * @param {Array} attributes The attributes array
+ * @return {*} True if the attribute value is valid, false otherwise
+ */
+function isValidAttributeValues(context, entityType, attributes) {
```
Comment (dmoranj) [inline lib/context_adapter_helper.js:166]: When the validations are that complex, it may be a good idea to split the function in multiple validation functions, one for each type of validation. This particular one is near the limit, I think. I leave it up to you if you think is appropriate, but, if you are planning to introduce more validations, I would split it to improve readability.

File: lib/context_adapter_helper.js [hunk 4]
```diff
@@ -333,12 +408,13 @@ function getOperationDescriptor(request, callback) {
...
  * Checks that an queryContext response returned by the Context Adapter to get
  *  service information is well-formed
+ * @param {object} context The context for logging
  * @param {object} response The response received from the Context Adapter
  * @param {string} serviceId The identifier of the service whose information was retrieved
  * @param {function} callback Callback to be called with the result of the check
  * @return {*} True if the queryContextResponse is valid. False otherwise.
  */
-function checkQueryContextResponse(response, serviceId, callback) {
+function checkQueryContextResponse(context, response, serviceId, callback) {
```
Comment (dmoranj) [inline lib/context_adapter_helper.js]: Why do you pass the logging context to the function? This is a bit confusing, and a bit dirty: I would limit the context to each module (to trace the logging of a request you can use domains), and I think the parameters of the function should reflect the information the function needs to do its work (and logging is a collateral task that shouldn't have a lot to do with the function).

File: lib/context_adapter_helper.js [hunk 5]
```diff
@@ -562,10 +667,21 @@ function getExternalRequestId(response) {
...
+      'Fiware-Service: ' + updateContextRequest.headers['fiware-service'] + ' and ' +
+      'Fiware-ServicePath: ' + updateContextRequest.headers['fiware-servicepath'] +
+      ' and payload: ' + JSON.stringify(payload)
+  );
+
   request({
     method: updateContextRequest.contextAdapter.serviceDescriptor.method,
-    uri: updateContextRequest.contextAdapter.serviceDescriptor.endpoint +
-          (updateContextRequest.contextAdapter.serviceDescriptor.mapping.path || ''),
+    uri: updateContextRequest.contextAdapter.serviceDescriptor.endpoint,
```
Comment (dmoranj) [inline lib/context_adapter_helper.js]: This literal object you are passing to request may be more readable if you define it in a variable above, instead of inlining the code inside the function call.

File: lib/context_adapter_helper.js (hunk 1 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: In fact, the way you suggest is the way it is done in the STH component :) And the same caLogger instance (once initialised) is passed to the rest of the modules of the app... This has the advantages of configuring the the logging just in one place and the disadvantage of making the JS files more dependant than it is really needed...

I guess you vote for the first approach, right? I mean, having the logging configured in just one place, right? :)

**Comment by dmoranj** [comment]
A general comment concerning the file `context_adapter_helper.js`: I think it is too big (800 lines or more is a trigger to think about splitting it), but, most importantly, I think it does lots of things related with the functionality that should belong to a file with a more descriptive name. Some examples are `notifyOperationClosed()`, `sendThirdPartyRequest()` or `sendUpdateContext()`. I would expect those to be in a `ngsiCommunications.js` file or a `thirdPartyComs.js` one or something like that. In a `XXX_Helper.js` I expect to find unimportant things like a function to encode or decode strings, or to split an array in a creative way, o something like that, but not the interpretation of the response of the Third Party (that is, the main features of the CA).

File: lib/context_adapter_helper.js (hunk 1 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: I mean, something such as in the "main file":

``` javascript
var caConfig = require('./context_adapter_configuration'),
    caLogger = require('logops');
caLogger.setLevel(caConfig.LOGOPS_LEVEL);

// Since these modules also need to log messages... The other option would be to include functions for logging in the main module and pass it to the other ones :p
var caHelper = require('./context_adapter_helper')(caLogger),
    caServer = require('./context_adapter_server')(caLogger);
```

File: lib/context_adapter_helper.js [hunk 6]
```diff
@@ -774,9 +871,8 @@ module.exports = {
   getServiceDescriptor: getServiceDescriptor,
   getOperationDescriptor: getOperationDescriptor,
   sendThirdPartyRequest: sendThirdPartyRequest,
-  notifyOperationInProgress: notifyOperationInProgress,
+  // notifyOperationInProgress: notifyOperationInProgress,
```
Comment (dmoranj) [inline lib/context_adapter_helper.js]: Never comment code. It either is useful or as to be erased.

File: lib/context_adapter_server.js [hunk 7]
```diff
@@ -25,6 +25,7 @@
 
 var caLogger = require('logops');
 var caConfig = require('./context_adapter_configuration');
+caLogger.setLevel(caConfig.LOGOPS_LEVEL);
```
Comment (dmoranj) [inline lib/context_adapter_server.js]: Same as above. And more interesting: do you realize that the same `setLevel()` function is being executed multiple times?

File: lib/context_adapter_helper.js (hunk 2 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Not really using it on the CA side but I do agree to include it since there are other attributes which are expected but not currently used... ;) Thanks, I'll include it in a sec.

File: lib/context_adapter_helper.js [hunk 8]
```diff
@@ -175,6 +177,16 @@ function sendQueryContext(service, servicePath, payload, callback) {
...
+    op: getOperationType()
+  };
+  caLogger.debug(
+    context,
+    'Sending updateContext to CB at ' + 'http://' + caConfig.CB_HOST + ':' + caConfig.CB_PORT + caConfig.CB_PATH +
+    '/updateContext' + ' with payload: ' + JSON.stringify(payload)
+  );
   request({
     method: 'POST',
     uri: 'http://' + caConfig.CB_HOST + ':' + caConfig.CB_PORT + caConfig.CB_PATH + '/updateContext',
```
Comment (mrutid) [inline lib/context_adapter_helper.js]: you may made just one string append

**Comment by dmoranj** [comment]
Concerning the log initialization, consider using the `start()` function in the `context_adapter_server.js` module.

File: lib/context_adapter_helper.js [hunk 9]
```diff
@@ -205,6 +217,71 @@ function isAttributeIncluded(attributes, attributeName) {
...
+  } else {
+    caLogger.debug(
+      context,
+      'Invalid entity type when validating attribute values: ' + entityType
+    );
+    return false;
+  }
+
+  for (var i = 0; i < attributes.length; i++) {
+    if (attributes[i].name === interactionTypeAttrName) {
```
Comment (mrutid) [inline lib/context_adapter_helper.js:184]: Is this ok? two ifs and no else branch... so... could they be merged into just one if sentence?

File: lib/context_adapter_helper.js (hunk 1 above)
Comment (dmoranj) [inline lib/context_adapter_helper.js]: I suggested a place below, but, reading your answer, I tend to think that you might have misunderstood how `require()` works. Did you know that modules are singletons? That the `require()` function executed n times creates a single object and returns it? You can leave the all the `require()` calls where they are and make a single centralized call to set the log level.

File: lib/context_adapter_helper.js [hunk 10]
```diff
@@ -516,43 +614,50 @@ function getServiceDescriptor(request, operationDescriptor, callback) {
...
-  return response;
+  var detailsObj;
+  var result = '';
+  if (data) {
+    if (data.body) {
+      detailsObj = data.body.details;
+    } else if (data.payload) {
+      detailsObj = data.payload.details;
+    } else {
+      return '0,' + 'XXX' + ',' + 'Unknown data received from the Third Party as the request result';
```
Comment (mrutid) [inline lib/context_adapter_helper.js]: why to append if there are no variables?

File: lib/context_adapter_helper.js [hunk 11]
```diff
@@ -573,94 +689,65 @@ function sendThirdPartyRequest(updateContextRequest, callback) {
...
+      ]
+    }
+  ];
+
+  if (request && request.contextAdapter.operationDescriptor.interactionType !==
+    caConfig.BUTTON_ENTITY.INTERACTION_TYPES.ASYNCHRONOUS) {
+    contextElements[0].attributes.push({
+      name: caConfig.BUTTON_ENTITY.IOTA_OPERATION_RESULT_ATTR_NAME,
+      type: caConfig.ATTRIBUTE_TYPE.STRING,
+      value: getResult(null, data)
```
Comment (mrutid) [inline lib/context_adapter_helper.js]: two calls, apparently equals, to getResult(null, data)

File: lib/context_adapter_helper.js (hunk 2 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/82779bf4cec59d461aeeff9e0c592fd4cb05c741

File: lib/context_adapter_helper.js (hunk 1 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Nice!!! :) So, I'm taking that path... :) Thanks!

File: lib/context_adapter_server.js [hunk 12]
```diff
@@ -363,15 +371,27 @@ function stop(callback) {
...
+    'Stopping the Context Adapter server...'
+  );
+
   if (server && server.info && server.info.started) {
     server.stop(function(err) {
       // Server successfully stopped
-      caLogger.info(context, 'hapi server successfully stopped');
+      caLogger.info(
+        context,
+        'hapi server successfully stopped'
```
Comment (mrutid) [inline lib/context_adapter_server.js]: consider something like "Http server (hapi) successfully stopped" since you are the only person using hapi

File: lib/context_adapter_server.js [hunk 13]
```diff
@@ -363,15 +371,27 @@ function stop(callback) {
...
+      );
+
       callback(err);
     });
   } else {
-    caLogger.info(context, 'No hapi server running');
+    caLogger.info(
+      context,
+      'No hapi server running'
```
Comment (mrutid) [inline lib/context_adapter_server.js]: same

File: lib/context_adapter_helper.js (hunk 1 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/e4eb9b44a5ba1749c70061332463f31fd3298869

File: lib/context_adapter_helper.js (hunk 3 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js:166]: If you are fine with it, I'll leave it as it is for the time being and as you mention if new validations are required we can take the approach you suggest :) Thanks!

File: lib/context_adapter_helper.js (hunk 4 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: The context depends on each request for better traceability (see https://github.com/telefonicaid/context-adapter/blob/task/synchronization-after-happy-path-e2e-integration/lib/context_adapter_server.js#L241) In fact, it depends on the UNICA_CORRELATOR header if present in the original request or a new one is generated for each specific request :)

I could not do it this way if you think that approach is the right one and obviously the context would be static at a JS file level with no need to pass it :)

Another approach to try to keep all the benefits and your suggestion would be to include the context in the response object passed via:

``` javascript
response.contextAdapter.context = theConcreteContextBasedOnEachRequest;
```

and I could take it from there. What do you think?

File: lib/context_adapter_helper.js (hunk 3 above)
Comment (dmoranj) [inline lib/context_adapter_helper.js:166]: Ok. NTC

File: lib/context_adapter_helper.js (hunk 5 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/4382bf3fe4ab7be061c0c0aa090f3b4f7bcada19

File: lib/context_adapter_helper.js (hunk 4 above)
Comment (dmoranj) [inline lib/context_adapter_helper.js]: Considering you are passing the response object already, it makes sense to do that. We use domains for the same process (but domains seem to have fallen in disgrace, so maybe the response object is a better place to set it). I consider the Correlator and the Context to be slightly different things, by the way (the correlator is for every request, and the context is the local, i.e.: module context).

File: lib/context_adapter_helper.js (hunk 6 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Ups, I missed that from some test :) Good catch! Thanks!

Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/b116d5f7db718931e087d47ee388ef3298211c99

File: lib/context_adapter_server.js (hunk 7 above)
Comment (gtorodelvalle) [inline lib/context_adapter_server.js]: Removed in a previous commit :) Thanks!

File: lib/context_adapter_helper.js (hunk 8 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: I'm sorry I do not get your point :( Could you rephrase it, please? :)

File: lib/context_adapter_helper.js (hunk 9 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js:184]: Thanks for the confirmation ;) Yeah, it is correct due to the use of the `return` statements ;)

File: lib/context_adapter_helper.js (hunk 10 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Good catch! :) I guess I was copy & pasting code... :D

Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/bb9e16a6164ba8bf30505511ef9e0568fc56812f

File: lib/context_adapter_helper.js (hunk 11 above)
Comment (gtorodelvalle) [inline lib/context_adapter_helper.js]: Almost identical :p Notice the value set to the name attribute. The reason is that in the synchronous button operation mode the `lazy_op_result` attribute should not be included and in the asynchronous button operation mode it should. That's the reason of that final `push()`.

File: lib/context_adapter_server.js (hunk 12 above)
Comment (gtorodelvalle) [inline lib/context_adapter_server.js]: I'm sexy and I know it... :p

Comment (gtorodelvalle) [inline lib/context_adapter_server.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/e12c0b89eded746437551829515ad75b8d39e6db Thanks!

File: lib/context_adapter_server.js (hunk 13 above)
Comment (gtorodelvalle) [inline lib/context_adapter_server.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/fde611d855715adabeadfedc6bbf1c8f83a4f933

**Comment by gtorodelvalle** [comment]
Regarding @dmoranj comment about splitting context_adapter_helper.js file into more meaningful ones, I do totally agree... ;) On it... New commit is coming... :p Thanks!

File: lib/context_adapter_helper.js (hunk 9 above)
Comment (mrutid) [inline lib/context_adapter_helper.js:184]: mmm... sure? if (L244) seems to do nothing special

File: lib/context_adapter_helper.js (hunk 11 above)
Comment (mrutid) [inline lib/context_adapter_helper.js]: I'm getting old... but... null is null... and data is the input parameter to the function... so... getResult(null, data) returns differently in  L731 and L720?

Comment (dmoranj) [inline lib/context_adapter_helper.js]: Marcos is right. What he says is you can store the result of the call to `getResult()` in a variable, then passing this variable to both object literals, instead of calling the function twice. It doesn't have anything to do with the push.

File: lib/context_adapter_helper.js (hunk 9 above)
Comment (dmoranj) [inline lib/context_adapter_helper.js:184]: I think Macos refers to the if in line 244. Considering there is no alternative branch neither in that if statement nor in the following one, both could be mixed using an ' &&'  operator.

**Comment by gtorodelvalle** [comment]
Extracting the communication related functions from `context_adapter_helper.js` into a new `context_adapter_communication.js` file fixed in https://github.com/telefonicaid/context-adapter/commit/3d9de01c051dbe0b0e228e22ba9e1bdc9e30b3ca

Really nice suggestion, BTW! Thanks @dmoranj

File: lib/context_adapter_communication.js [hunk 14]
```diff
@@ -0,0 +1,677 @@
...
+                      required: true,
+                      pattern: 'false'
+                    },
+                    attributes: {
+                      id: 'attributes',
+                      type: 'array',
+                      required: true,
+                      minItems: 6,
+                      maxItems: 8,
+                      conform: function(attributes) {
```
Comment (dmoranj) [inline lib/context_adapter_communication.js]: Extract this function to an external function, it seems weird to define a synchronous function inside a literal object declaration.

File: lib/context_adapter_communication.js [hunk 15]
```diff
@@ -0,0 +1,677 @@
...
+                  required: true,
+                  pattern: 'false'
+                },
+                attributes: {
+                  id: 'attributes',
+                  type: 'array',
+                  required: true,
+                  minItems: 4,
+                  maxItems: 6,
+                  conform: function(attributes) {
```
Comment (dmoranj) [inline lib/context_adapter_communication.js]: Same thing as below: extract this function to an external function.

File: lib/context_adapter_communication.js (hunk 14 above)
Comment (gtorodelvalle) [inline lib/context_adapter_communication.js]: Agreed! ;) On it... ;)

Comment (gtorodelvalle) [inline lib/context_adapter_communication.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/4e6c67511f36722f78cc717a7c5ae97b88c5ab54

File: lib/context_adapter_communication.js (hunk 15 above)
Comment (gtorodelvalle) [inline lib/context_adapter_communication.js]: Fixed in https://github.com/telefonicaid/context-adapter/commit/4e6c67511f36722f78cc717a7c5ae97b88c5ab54
